rust/main/chains/hyperlane-aleo/src/indexer/merkle_tree_hook.rs

```diff
@@ -0,0 +1,206 @@
+        Box::new(self.client.clone())
+    }
+}
+
+impl HyperlaneContract for AleoMerkleTreeHook {
+    /// Address
+    fn address(&self) -> H256 {
+        self.address
+    }
+}
+
+impl AleoIndexer for AleoMerkleTreeHook {
+    const INDEX_MAPPING: &str = "last_event_index";
+    const VALUE_MAPPING: &str = "inserted_into_tree_events";
+
+    type AleoType = InsertIntoTreeEvent;
+    type Type = MerkleTreeInsertion;
+
+    fn get_client(&self) -> &AleoProvider {
+        &self.client
+    }
+
+    fn get_program(&self) -> &str {
+        &self.program
+    }
+
+    /// Returns the latest event index of that specific block
+    async fn get_latest_event_index(&self, height: u32) -> ChainResult<u32> {
+        let key = HookEventIndex {
+            hook: self.aleo_address,
+            block_height: height,
+        };
+        // The latest event index for hooks is composition of block_height & hook_address
+        let last_event_index: u32 = self
+            .get_client()
+            .get_mapping_value(self.get_program(), Self::INDEX_MAPPING, &key)
+            .await?;
+        Ok(last_event_index)
+    }
+
+    /// Returns the event value of a mapping
+    fn get_mapping_key<N: Network>(&self, index: u32) -> ChainResult<Plaintext<N>> {
+        let str_value = format!("{{hook: {}, index: {}u32}}", self.aleo_address, index);
+        Ok(Plaintext::from_str(&str_value).map_err(HyperlaneAleoError::from)?)
+    }
+}
+
+#[async_trait]
+impl Indexer<MerkleTreeInsertion> for AleoMerkleTreeHook {
+    /// Fetch list of logs between blocks `from` and `to`, inclusive.
+    async fn fetch_logs_in_range(
+        &self,
+        range: RangeInclusive<u32>,
+    ) -> ChainResult<Vec<(Indexed<MerkleTreeInsertion>, LogMeta)>> {
+        AleoIndexer::fetch_logs_in_range(self, range).await
+    }
+
+    /// Get the chain's latest block number that has reached finality
+    async fn get_finalized_block_number(&self) -> ChainResult<u32> {
+        AleoIndexer::get_finalized_block_number(self).await
+    }
+
+    /// Fetch list of logs emitted in a transaction with the given hash.
+    async fn fetch_logs_by_tx_hash(
+        &self,
+        tx_hash: H512,
+    ) -> ChainResult<Vec<(Indexed<MerkleTreeInsertion>, LogMeta)>> {
+        AleoIndexer::fetch_logs_by_tx_hash(self, tx_hash).await
+    }
+}
+
+#[async_trait]
+impl SequenceAwareIndexer<MerkleTreeInsertion> for AleoMerkleTreeHook {
+    /// Return the latest finalized sequence (if any) and block number
+    async fn latest_sequence_count_and_tip(&self) -> ChainResult<(Option<u32>, u32)> {
+        let (mth, height) = self
+            .client
+            .get_mapping_value_meta::<AleoMerkleTreeHookStruct>(
+                &self.program,
+                "merkle_tree_hooks",
+                &self.aleo_address.to_string(),
+            )
+            .await?;
+        Ok((Some(mth.tree.count), height))
+    }
+}
+
+#[async_trait]
+impl MerkleTreeHook for AleoMerkleTreeHook {
```

Comment:
It looks like we are not relying on `reorg_period` and block height in queries. Why don't we do it?

Reply:
I cannot find the comment.

Reply:
I found doc comment to method using `reorgPeriod`.